dm/master/openapi_test.go

```diff
@@ -206,4 +220,131 @@ func (t *openAPISuite) TestSourceAPI(c *check.C) {
 	c.Assert(resultListSource2.Total, check.Equals, 0)
+	cancel()
+}
+
+func (t *openAPISuite) TestRelayAPI(c *check.C) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ctrl := gomock.NewController(c)
+	defer ctrl.Finish()
+	s := setupServer(ctx, c)
+	defer s.Close()
+
+	baseURL := "/api/v1/sources"
+
+	source1 := openapi.Source{
+		SourceName: source1Name,
+		EnableGtid: false,
+		Host:       "127.0.0.1",
+		Password:   "123456",
+		Port:       3306,
+		User:       "root",
+	}
+	result := testutil.NewRequest().Post(baseURL).WithJsonBody(source1).Go(t.testT, s.echo)
+	// check http status code
+	c.Assert(result.Code(), check.Equals, http.StatusCreated)
+
+	// get source status
+	source1StatusURL := fmt.Sprintf("%s/%s/status", baseURL, source1.SourceName)
+	result2 := testutil.NewRequest().Get(source1StatusURL).Go(t.testT, s.echo)
+	c.Assert(result2.Code(), check.Equals, http.StatusOK)
+
+	var getSourceStatusResponse openapi.GetSourceStatusResponse
+	err := result2.UnmarshalBodyToObject(&getSourceStatusResponse)
+	c.Assert(err, check.IsNil)
+	c.Assert(getSourceStatusResponse.Data[0].SourceName, check.Equals, source1.SourceName)
+	c.Assert(getSourceStatusResponse.Data[0].WorkerName, check.Equals, "") // no worker bound
+	c.Assert(getSourceStatusResponse.Total, check.Equals, 1)
+
+	// add mock worker the unbounded sources should be bounded
+	ctx1, cancel1 := context.WithCancel(ctx)
+	defer cancel1()
+	workerName1 := "worker1"
+	c.Assert(s.scheduler.AddWorker(workerName1, "172.16.10.72:8262"), check.IsNil)
+	go func(ctx context.Context, workerName string) {
+		c.Assert(ha.KeepAlive(ctx, s.etcdClient, workerName, keepAliveTTL), check.IsNil)
+	}(ctx1, workerName1)
+	// wait worker ready
+	c.Assert(utils.WaitSomething(30, 100*time.Millisecond, func() bool {
+		w := s.scheduler.GetWorkerBySource(source1.SourceName)
+		return w != nil
+	}), check.IsTrue)
+
+	// mock worker get status relay not started
+	mockWorkerClient := pbmock.NewMockWorkerClient(ctrl)
+	mockRelayQueryStatus(mockWorkerClient, source1.SourceName, workerName1, pb.Stage_InvalidStage)
+	s.scheduler.SetWorkerClientForTest(workerName1, newMockRPCClient(mockWorkerClient))
+
+	// get source status again,source should be bounded by worker1,but relay not started
+	result3 := testutil.NewRequest().Get(source1StatusURL).Go(t.testT, s.echo)
+	c.Assert(result3.Code(), check.Equals, http.StatusOK)
+	var getSourceStatusResponse2 openapi.GetSourceStatusResponse
+	err = result3.UnmarshalBodyToObject(&getSourceStatusResponse2)
+	c.Assert(err, check.IsNil)
+	c.Assert(getSourceStatusResponse2.Data[0].SourceName, check.Equals, source1.SourceName)
+	c.Assert(getSourceStatusResponse2.Data[0].WorkerName, check.Equals, workerName1) // worker1 is bound
+	c.Assert(getSourceStatusResponse2.Data[0].RelayStatus, check.IsNil)              // not start relay
+	c.Assert(getSourceStatusResponse2.Total, check.Equals, 1)                        // no worker bound
+
+	// start relay
```

Comment:
seems you didn't test start relay, for example, it has written something to etcd.